db/snapshot/backup_task.cc

```diff
@@ -99,52 +101,133 @@ future<> backup_task_impl::upload_component(sstring name) {
     }
 }
 
-future<> backup_task_impl::do_backup() {
-    if (!co_await file_exists(_snapshot_dir.native())) {
-        throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
+namespace {
+
+struct shard_files {
+    size_t size{0};
+    std::vector<sstring> files;
+    std::strong_ordering operator<=>(const shard_files& other) const noexcept { return size <=> other.size; }
+};
+
+using sharded_files = std::vector<shard_files>;
+
+sharded_files greedy_partitioning(std::vector<std::tuple<size_t, sstring>>&& files, size_t partitions) {
+    sharded_files result(partitions);
+    std::ranges::sort(files, std::greater());
+    for (auto&& [size, file] : files) {
+        auto& smallest = *std::ranges::min_element(result, std::less());
+        smallest.files.emplace_back(std::move(file));
+        smallest.size += size;
+    }
+    return result;
+}
+
+[[maybe_unused]] sharded_files round_robin_partitioning(std::vector<std::tuple<size_t, sstring>>&& files, size_t partitions) {
+    sharded_files result(partitions);
+    std::ranges::sort(files, std::greater());
+    size_t pos = 0;
+    for (auto&& [size, file] : files) {
+        auto& elem = result[pos++ % partitions];
+        elem.files.emplace_back(std::move(file));
+        elem.size += size;
+    }
+    return result;
+}
+
+sstring print_results(const sharded_files& files) {
+    sstring sizes_table;
+    sizes_table += "\n+-------+----------------------------------------------+\n";
+    sizes_table += "|  Shard  |    Accumulated size    |  Number of files  |\n";
+    sizes_table += "+---------+--------------------------------------------+\n";
+    for (size_t i = 0; i < files.size(); ++i) {
+        sizes_table += std::format("|{:^9}|{:>17} bytes |{:>19}|\n", i, files[i].size, files[i].files.size());
     }
+    sizes_table += "+---------+------------------------+-------------------+\n";
+    return sizes_table;
+}
 
+future<sharded_files> get_backup_files(const std::filesystem::path& snapshot_dir) {
     std::exception_ptr ex;
-    gate uploads;
-    auto snapshot_dir_lister = directory_lister(_snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+    auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
```

Comment:
The only called of get_backup_files() passes it `_snapshot_dir` as argument, so why can't get_backup_files() continue using it as it did before the patch?

Reply:
Since it's not a member function, I don't see any benefit in converting it to one. However, I am open to making it a member function if needed.

Reply:
A member function would be better IMO